### ISSUE TITLE
Weird horizontal stripe in Australian catchment

### ISSUE THREAD

**Issue by cjacobscrisioni** [ObjectVision/NetworkModel_World#18]

**Comment by cjacobscrisioni** [comment]
![image](https://github.com/ObjectVision/NetworkModel_World/assets/144823297/43fcce88-0895-4ad9-92b0-8eea243355ed)

**Comment by cjacobscrisioni** [comment]
Problem seems to come into play when converting lat/lon to webmercator yields a null value... See eg id= 741842 of NetworkSetup/GridPoints_Per_Continent/Australia_Oceania/select/centroid_wm.

See table below of a grid cell missing in the above stripe.
 
![image](https://github.com/ObjectVision/NetworkModel_World/assets/144823297/f29227ae-e674-4722-b70a-f39c34e6a488)

Code=attribute<WebMercator> centroid_wm := convert(point(PointCol(centroid), PointRow(centroid))[WGS84], Geometries/WebMercator);

**Comment by cjacobscrisioni** [comment]
Short list of problematic grid centroids from /Analyses/Create_Centres_Catchments/Australia_Oceania/Country_Catchments/AUS/Country_Unconnected_Grids/geometry

"id";"geometry_LL";"geometry"
124154;{-14.2011627, 143.652459};{null, null}
124155;{-14.2011627, 143.66263};{null, null}
124156;{-14.2011627, 143.6728};{null, null}
124157;{-14.2011627, 143.682971};{null, null}
124158;{-14.2011627, 143.693142};{null, null}
124328;{-14.209347, 143.645497};{null, null}
124329;{-14.209347, 143.655668};{null, null}
124330;{-14.209347, 143.665839};{null, null}
124331;{-14.209347, 143.676011};{null, null}
124332;{-14.209347, 143.686182};{null, null}
124333;{-14.209347, 143.696353};{null, null}
124505;{-14.2175315, 143.648709};{null, null}
124506;{-14.2175315, 143.65888};{null, null}
124507;{-14.2175315, 143.669051};{null, null}
124508;{-14.2175315, 143.679223};{null, null}
124509;{-14.2175315, 143.689394};{null, null}
124510;{-14.2175315, 143.699565};{null, null}

**Comment by cjacobscrisioni** [comment]
@jipclaassens , @MaartenHilferink , before I proceed to further reproduce this issue - I get null values when converting a limited set of lat/lon coordinates to webmercator. These manifest as a horizontal bar in Australia. Happens on two computers. The problem does not seem to occur northwards, southwards, westwards or eastwards of the bar. Any ideas where to look?

**Comment by jipclaassens** [comment]
It already goes wrong here: 
`/NetworkSetup/GridPoints_Per_Continent/Australia_Oceania/select/centroid_wm`

@eoudejans is working on a fix for the conversion error we are seeing. Discussed in this issue (https://github.com/ObjectVision/GeoDMS/issues/563). We will let you know when we know more. It is a high priority at the moment.

**Comment by jipclaassens** [comment]
![image](https://github.com/ObjectVision/NetworkModel_World/assets/48675272/fe94c197-ea1d-4079-8dcf-1e78cc58d0a6)

**Comment by MaartenHilferink** [comment]
Alternative impl: https://gis.stackexchange.com/questions/177134/transforming-lat-and-long-to-mercator-x-and-y

**Comment by jipclaassens** [comment]
heeft geen null waarden:
`/SourceData/population/sel_LandCells/org_rel_to_mollweide `

deze wel:
`/SourceData/population/sel_LandCells/latlon_point      := convert( [org_rel_to_mollweide ]

**Comment by eoudejans** [comment]
Using the following code selects only points that are null after conversion to mollweide but defined in EPSG:4326.

```
unit<uint32> unique_nulls_in_wm := select_with_attr_by_org_rel(inGrids_centroid_wgs, !IsNull(inGrids_centroid_wgs) && IsNull(inGrids_centroid_wm))
{
	attribute<BaseProjection> points := inGrids_centroid_wgs[org_rel];
}
```

Results in the following points:

![image](https://github.com/ObjectVision/GeoDMS/assets/96182097/49095a0c-abc4-4d5e-904d-c0e0712fdd17)

Left hemisphere Islands:
![image](https://github.com/ObjectVision/GeoDMS/assets/96182097/a600271c-874e-4d46-acde-753e2d783f33)

Right hemisphere:
![image](https://github.com/ObjectVision/GeoDMS/assets/96182097/26de8e7b-4df7-4507-b6f7-ed9e72e0c121)

**Comment by eoudejans** [comment]
Changing the blocksize we use that goes into the transformation using PROJ9 from 1024 to 32 resulted in 64 missing values down from 2048. And these remaining points are closer to the datum crossover:
![image](https://github.com/ObjectVision/NetworkModel_World/assets/96182097/fcd64f77-7b6d-4be1-947b-c3a13f69cd6d)

This leads us to the solution of recursively downsizing the block size that goes into the transformation by half from 1024 down to 1 or until no more nulls are present.

**Comment by eoudejans** [comment]
Latest commit fixed the unconverted points, see overview:
![image](https://github.com/ObjectVision/NetworkModel_World/assets/96182097/1a3c24e9-ba9d-4b7a-a7eb-89f46597ea28)

Zoom-in on previously untransformed points:
![image](https://github.com/ObjectVision/NetworkModel_World/assets/96182097/170e4055-b887-4825-85d4-ca600331c368)

**Comment by eoudejans** [comment]
Further confirmation that there no longer exist null-values in the transformed web mercator points that do not exist in the wgs point set for Oceania:
![image](https://github.com/ObjectVision/NetworkModel_World/assets/96182097/b22cd06c-8b8c-4d1b-bcad-a830e5b94037)

**Comment by cjacobscrisioni** [comment]
This is awesome!!!! Thanks all :) 

Two questions -> 
1) this requires an upcoming new version of GeoDMS, right? and 
2) before recursively downsizing the block size, do you verify whether the original lat/lon coordinates are within lat/lon bounds? Thinking of the rare case that block sizes are incrementally reduced because of impossible input coordinates.

**Comment by eoudejans** [comment]
> 1. this requires an upcoming new version of GeoDMS, right?
Yes, 14.9.0 will be published as pre-release, likely today. (we pre-released it alteady but due to remaining issues related to this issue retracted it)

> 2. before recursively downsizing the block size, do you verify whether the original lat/lon coordinates are within lat/lon bounds? Thinking of the rare case that block sizes are incrementally reduced because of impossible input coordinates.
This could be done, but requires potentially expensive bookkeeping to make sure only blocks of arbitrarily nearby points go in the transformation as one block.

Recursively downsizing the blocksize goes down to reprojecting per 1 coordinate, which will not cross the projection bounds as that requires the minimum of two points.

**Comment by jipclaassens** [comment]
https://github.com/ObjectVision/GeoDMS/issues/563#issuecomment-1886709873

Does this mean we must omit all swapping codes with the new direct code?

**Comment by eoudejans** [comment]
> [ObjectVision/GeoDMS#563 (comment)](https://github.com/ObjectVision/GeoDMS/issues/563#issuecomment-1886709873)
> 
> Does this mean we must omit all swapping codes with the new direct code?

At least all the workarounds I have tested no longer work, this involved the cases you (@jipclaassens) created in https://github.com/ObjectVision/GeoDMS/issues/563 and the case Bas created from which https://github.com/ObjectVision/GeoDMS/issues/563 followed.

I assume that from 14.9.0 swapping the workarounds to the direct conversion is a must.
